StereoKitC/tools/file_picker.cpp

```diff
@@ -435,30 +467,118 @@ void file_picker_update() {
 			if (fp_active == nullptr) ui_push_enabled(false);
 			if (ui_button("Open")) { snprintf(fp_filename, sizeof(fp_filename), "%s%c%s", fp_path.folder, platform_path_separator_c, fp_active); fp_call = true; fp_call_status = true; }
 			ui_sameline();
+			ui_push_enabled(true);
+			if (ui_button("ListMode")) {
+				fp_list_mode = !fp_list_mode;
+			}
+			ui_sameline();
+			if (fp_active == nullptr) ui_push_enabled(false);
 			ui_label(fp_active?fp_active:"None selected...");
 			if (fp_active == nullptr) ui_pop_enabled();
 		} break;
 		}
 
 		ui_hseparator();
 
+		if (spr_folder == nullptr) {
+			spr_folder = sprite_find(ui_default_id_spr_arrow_right);
+		}
+		if (spr_up == nullptr) {
+			spr_up = sprite_find(ui_default_id_spr_arrow_up);
+		}
+		static ui_btn_layout_ ui_folderimg_button_layout = ui_btn_layout_left;
 		// List the files
 		vec2 size = { .12f, line_height * 1.5f };
+		if (fp_list_mode) {
+			static ui_btn_layout_ layout = ui_btn_layout_left;
+			if (fp_sortby == Name) {
+				if (ui_button_img_sz("Name", spr_up, layout, size)) {
+					fp_sortby = Name;
+					fp_sort_order_changed = true;
+				}
+			}
+			else {
+				if (ui_button_sz("Name", size)) {
+					fp_sortby = Name;
+					fp_sort_order_changed = true;
+				}
+			}
+			ui_sameline();
+			ui_layout_reserve(size, false, 0.0f);
+			ui_sameline();
+			if (fp_sortby == Size) {
+				if (ui_button_img_sz("Size", spr_up, layout, size)) {
+					fp_sortby = Size;
+					fp_sort_order_changed = true;
+				}
+			}
+			else {
+				if (ui_button_sz("Size", size)) {
+					fp_sortby = Size;
+					fp_sort_order_changed = true;
+				}
+			}
+			ui_hseparator();
+			if (fp_sort_order_changed) {
+				switch (fp_sortby) {
+				case Name:
+					fp_items.sort([](const fp_item_t& a, const fp_item_t& b) { return a.file != b.file ? a.file - b.file : strcmp(a.name, b.name); });
+					break;
+				case Size:
+					fp_items.sort([](const fp_item_t& a, const fp_item_t& b) { return a.size == b.size ? 0 : a.size - b.size > 0 ? 1 : -1;});
+					break;
+				}
+			}
+		}
 		const int32_t scroll_cols = 3;
 		const int32_t scroll_rows = 5;
-		const int32_t scroll_step = scroll_cols*scroll_rows;
+		const int32_t scroll_step = fp_list_mode ? scroll_rows : scroll_cols * scroll_rows;
 		vec3 file_grid_start = ui_layout_at();
 		ui_panel_begin();
 		for (int32_t i = fp_scroll_offset; i < fp_scroll_offset + scroll_step; i++) {
-			if (i >= fp_items.count) {
-				ui_layout_reserve(size);
-			} else if (ui_button_sz(fp_items[i].name, size)) {
-				if (fp_items[i].file)
-					fp_active = fp_items[i].name;
+			if (fp_list_mode) {
+				if (i >= fp_items.count) {
+					ui_layout_reserve(size, false, 0.0f);
+					ui_sameline();
+					ui_layout_reserve(size, false, 0.0f);
+					ui_sameline();
+					ui_layout_reserve(size, false, 0.0f);
```

Comment:
For cleanliness sake, I'd prefer just one call to ui_layout_remaining with the correct size, instead of 3.

Reply:
Tried and tested with ui_layout_remaining. It is not able to correctly place the scroll down button. The original Non-list mode assumes 5 rows of data and places the scroll down button. To keep in parity with the same design choice in list mode, I am adding empty space to accommodate all the rows so that, the scroll down button is in sync with the panel